messaging/src/main/java/org/axonframework/common/infra/FilesystemStyleComponentDescriptor.java

```diff
@@ -0,0 +1,377 @@
+
+    private static final String ROOT_PATH = "/";
+    private static final String PATH_SEPARATOR = "/";
+
+    private final Map<Object, String> componentPaths;
+    private final Map<String, Object> properties;
+    private final String currentPath;
+
+    /**
+     * Constructs a new {@code FilesystemComponentDescriptor} as the root of the hierarchy.
+     */
+    public FilesystemStyleComponentDescriptor() {
+        this(new IdentityHashMap<>(), ROOT_PATH);
+    }
+
+    /**
+     * Private constructor used for creating nested descriptors that share the component paths map.
+     *
+     * @param componentPaths Map containing paths to already processed components.
+     * @param currentPath    The current path in the virtual filesystem.
+     */
+    private FilesystemStyleComponentDescriptor(Map<Object, String> componentPaths, String currentPath) {
+        this.componentPaths = componentPaths;
+        this.currentPath = currentPath;
+        this.properties = new LinkedHashMap<>();
+    }
+
+    @Override
+    public void describeProperty(@Nonnull String name, Object object) {
+        if (object instanceof DescribableComponent component) {
+            describeComponent(name, component);
+        } else {
+            properties.put(name, object);
+        }
+    }
+
+    private void describeComponent(String name, DescribableComponent component) {
+        var componentPath = currentPathChild(name);
+
+        boolean componentSeenAlready = componentPaths.containsKey(component);
+        if (componentSeenAlready) {
+            var existingPath = componentPaths.get(component);
+            properties.put(name, new SymbolicLink(existingPath));
+            return;
+        }
+
+        componentPaths.put(component, componentPath);
+        properties.put(name, componentDescriptor(component, componentPath));
+    }
+
+    private String currentPathChild(String name) {
+        return currentPath.equals(ROOT_PATH)
+                ? currentPath + name
+                : currentPath + PATH_SEPARATOR + name;
+    }
+
+    @Override
+    public void describeProperty(@Nonnull String name, Collection<?> collection) {
+        if (collection == null) {
+            properties.put(name, null);
+            return;
+        }
+        var items = new ArrayList<>();
+
+        int index = 0;
+        for (var item : collection) {
+            var property = item instanceof DescribableComponent component
+                    ? describeComponentInCollection(name, index, component)
+                    : item;
+            items.add(property);
+            index++;
+        }
+
+        properties.put(name, items);
+    }
+
+    private Object describeComponentInCollection(
+            String name,
+            int index,
+            DescribableComponent component
+    ) {
+        var itemName = name + "[" + index + "]";
+        var itemPath = currentPathChild(itemName);
```

Comment:
Super nit: Similar pointer as earlier. This value is used in the else-block, but constructed upfront. As such, it might not be used at all when `componentSeenAlready` is true.